### PR TITLE
feat: add turnkey signer to aa-signers

### PR DESCRIPTION
**Pull request by oliviathet**
# Pull Request Checklist

- [x] Did you add new tests and confirm existing tests pass? (`yarn test`)
- [x] Did you update relevant docs? (docs are found in the `site` folder, and guidelines for updating/adding docs can be found in the [contribution guide](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md))
- [x] Do your commits follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] Does your PR title also follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) standard?
- [x] If you have a breaking change, is it [correctly reflected in your commit message](https://www.conventionalcommits.org/en/v1.0.0/#examples)? (e.g. `feat!: breaking change`)
- [x] Did you run lint (`yarn lint:check`) and fix any issues? (`yarn lint:fix`)
- [x] Is the base branch you're merging into `development` and not `main`?
- [x] Did you follow the [contribution guidelines](https://github.com/alchemyplatform/aa-sdk/blob/main/CONTRIBUTING.md)?

---

## PR-Codex overview
This PR adds support for the `TurnkeySigner` in the `aa-signers` package. 

### Detailed summary
- Export `TurnkeySigner` and `TurnkeyAuthParams` from `@alchemy/aa-signers`
- Export `TurnkeySigner` and `TurnkeySubOrganization` from `@alchemy/aa-signers/turnkey`
- Add types and classes for `TurnkeyAuthParams`, `TurnkeySubOrganization`, and `TurnkeyUserMetadata`
- Update `package.json` with new dev dependencies
- Add documentation for `getAddress`, `signMessage`, `getAuthDetails`, and `constructor` methods in `TurnkeySigner`

> The following files were skipped due to too many changes: `site/snippets/turnkey.ts`, `packages/signers/src/turnkey/signer.ts`, `site/packages/aa-signers/turnkey/introduction.md`, `packages/signers/src/turnkey/__tests__/signer.test.ts`, `site/smart-accounts/signers/turnkey.md`, `site/packages/aa-signers/turnkey/authenticate.md`, `yarn.lock`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (moldy530) [inline .github/PULL_REQUEST_TEMPLATE.md:8]: thank you

File: packages/signers/package.json
Comment (moldy530) [inline packages/signers/package.json:72]: is `@turnkey/api-key-stamper` not required here too?

File: packages/signers/src/turnkey/types.ts
Comment (moldy530) [inline packages/signers/src/turnkey/types.ts:6]: is this an `Address`?

File: site/snippets/turnkey.ts
Comment (moldy530) [inline site/snippets/turnkey.ts:19]: nit: should this be:
```suggestion
    // API Key, WebAuthn, or Email Auth [stampers](https://docs.turnkey.com/category/api-design)
```

**Review by moldy530 — COMMENTED** [review]
this is awesome @oliviathet! thanks for doing this. had a few comments and clarifying questions

File: site/packages/aa-signers/turnkey/introduction.md
Comment (oliviathet) [reply site/packages/aa-signers/turnkey/introduction.md:53]: definitely! great point

File: packages/signers/package.json
Comment (oliviathet) [reply packages/signers/package.json:72]: the stamper packages are only used in the examples and tests, which is why i ultimately opted to place it under `devDependencies` but not in `optionalDependencies`. it's up to the application whether they want to use the webauthn, api key, or email auth stamper. even though these stamper dependencies aren't required for this package, i'll defer to you all as to whether it makes the developer experience smoother to add it here.

**Comment by moldy530** [comment]
Quick update, we're switching back to using `main` as our default branch, could you rebase this PR on top of main? (i really hope this doesn't cause you a massive rebase headache >.>) if you run into issues lemme know

File: packages/signers/src/turnkey/types.ts
Comment (oliviathet) [reply packages/signers/src/turnkey/types.ts:6]: an `Address` or `ID`. i also describe it [here](https://github.com/alchemyplatform/aa-sdk/pull/292/files#diff-fdf51fa5129147a3ef7bf55922343105cd7e6f7d4995e147c019e81662794ca7R67) but lmk if some additional comments anywhere would be clarifying.

File: site/packages/aa-signers/turnkey/introduction.md
Comment (oliviathet) [reply site/packages/aa-signers/turnkey/introduction.md:65]: the example doesn't create the wallet, so i agree that it might be helpful to show how the dev can obtain the wallet id/address f they don't store that for users. since the wallet id/address is only needed for the `authenticate` call and this alternative example is a bit lengthy, wdyt about me adding it [here](https://github.com/alchemyplatform/aa-sdk/pull/292/files#diff-fdf51fa5129147a3ef7bf55922343105cd7e6f7d4995e147c019e81662794ca7R25)?

Comment (moldy530) [reply site/packages/aa-signers/turnkey/introduction.md:65]: yea that sounds good to me

File: site/packages/aa-signers/turnkey/authenticate.md
Comment (moldy530) [inline site/packages/aa-signers/turnkey/authenticate.md:45]: for a new user this will always be true right? maybe let's add a call to `createWallet` or `createSubOrg` here so we can generate a new user / wallet

**Review by moldy530 — COMMENTED** [review]
hype!! just had one more nit on the example but LGTM otherwise

File: site/smart-accounts/signers/turnkey.md
Comment (avasisht23) [inline site/smart-accounts/signers/turnkey.md:35]: since the snippet uses `@turnkey/webauthn-stamper` here, I'd recommend mentioning a recommendation here re: webauthn / the alternatives you mentioned above (webauthn, api key, or email auth stamper)

File: packages/signers/src/turnkey/types.ts
Comment (avasisht23) [inline packages/signers/src/turnkey/types.ts:4]: out of curiosity, are these types and/or fields in some base/core JS SDK that Turnkey has that can be imported? would love to not have to maintain these if there's one source of truth :)

File: site/packages/aa-signers/turnkey/authenticate.md
Comment (avasisht23) [reply site/packages/aa-signers/turnkey/authenticate.md:45]: by here you mean as an object on `authParams` right? agree, would love this to be part of the `authenticate` flow.

**Review by avasisht23 — COMMENTED** [review]
LGTM after last few comments! will stamp when revised. Thanks Olivia!

File: site/packages/aa-signers/turnkey/authenticate.md
Comment (oliviathet) [reply site/packages/aa-signers/turnkey/authenticate.md]: this is a great call out @avasisht23 ! i wrote the code with the assumption that between initializing `turnkeyClient` and `turnkeySigner`, the dev creates or retrieves a sub-organization and wallet for the user. but as @moldy530 refers to in the first comment, there's multiple flows, including:
- already have sub-organization id and wallet -> `getWallets`/`getWallet`
- already have sub-organization -> `createWallet`
- `createSubOrganization`

to capture this flow as a method in `authParams` that the dev can implement, i've replaced `organizationId` and `signWith` with `resolveSubOrganization()`. i'm going to be honest -- the turnkey sdk could really be improved to simplify the dev's current implementation. fortunately, we're working on some design changes to the turnkey sdk. for example, these `create` endpoints will become synchronous and their new response type can be used instead of maintaining a `TurnkeySubOrganization` type in turnkey/types. (related to: https://github.com/alchemyplatform/aa-sdk/pull/292#discussion_r1414216980)

File: packages/signers/src/turnkey/types.ts
Comment (oliviathet) [reply packages/signers/src/turnkey/types.ts:4]: currently no. but i totally agree and we may be making some changes here in the near future by importing a type from the turnkey sdk. more in this comment: https://github.com/alchemyplatform/aa-sdk/pull/292#discussion_r1415978792

**Review by moldy530 — APPROVED** [review]
nice LGTM! I'll grab this and handle the rebase

**Comment by moldy530** [comment]
closing in favor of: https://github.com/alchemyplatform/aa-sdk/pull/307

**Pull request closed without merging**
